src/main/java/frc/robot/subsystems/vision/Vision.java

```diff
@@ -0,0 +1,95 @@
+import edu.wpi.first.wpilibj.DriverStation;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import frc.robot.Constants.VisionConstants;
+import frc.robot.Constants.FieldConstants;;
+
+/**
+ * Runnable that gets AprilTag data from PhotonVision.
+ */
+public class Vision implements Runnable {
+
+  private final PhotonPoseEstimator photonPoseEstimator;
+  private final PhotonCamera photonCamera;
+  private final AtomicReference<EstimatedRobotPose> atomicEstimatedRobotPose = new AtomicReference<EstimatedRobotPose>();
+
+  public Vision(PhotonCamera cameraName, Transform3d robotToCamera) {
+    this.photonCamera = cameraName;
+    PhotonPoseEstimator photonPoseEstimator = null;
+
+    try {
+      var layout = AprilTagFields.k2024Crescendo.loadAprilTagLayoutField();
+      layout.setOrigin(OriginPosition.kBlueAllianceWallRightSide); // Sets the Origin to the Blue AllianceWall and will be flipped in the robot thread
+
+      // if (photonCamera != null) {
+      photonPoseEstimator = photonCamera != null
+          ? new PhotonPoseEstimator(layout, PoseStrategy.MULTI_TAG_PNP_ON_COPROCESSOR, photonCamera, robotToCamera)
+          : null;
+      photonPoseEstimator.setMultiTagFallbackStrategy(PoseStrategy.LOWEST_AMBIGUITY);
+      // }
+    } catch (Exception e) {
+      DriverStation.reportError("Failed to load AprilTagFieldLayout", e.getStackTrace());
+      // photonPoseEstimator = null;
+    }
+
+    this.photonPoseEstimator = photonPoseEstimator;
+  }
+
+  @Override
+  public void run() {
+    // Get AprilTag data and updating the pose estimator
+    try {
+      if (photonPoseEstimator != null && photonCamera != null) {
+        var photonResults = photonCamera.getLatestResult(); //Gets the latest camera results
+        SmartDashboard.putString(photonCamera.getName(),
+            photonCamera.getLatestResult().getMultiTagResult().estimatedPose.toString());
+        if (photonResults.hasTargets() && (photonResults.targets.size() > 1
+            || photonResults.targets.get(0).getPoseAmbiguity() < VisionConstants.APRILTAG_AMBIGUITY_THRESHOLD)) {
+          //Updates the pose estimator
+          photonPoseEstimator.update(photonResults).ifPresent(estimatedRobotPose -> {
+            var estimatedPose = estimatedRobotPose.estimatedPose;
+            // SmartDashboard.putNumber("estimatedPoseX", estimatedPose.toPose2d().getX());
+            // SmartDashboard.putNumber("estimatedPoseY", estimatedPose.toPose2d().getY());
+            /** 
+            * If present then makes sure the measurement is on the field and
+            * sets the atomic estimated pose to the current estimated pose
+            * from PhotonPoseEstimator
+            */
+            if (estimatedPose.getX() > 0.0 && estimatedPose.getX() <= FieldConstants.length
+                && estimatedPose.getY() > 0.0
+                && estimatedPose.getY() <= FieldConstants.width) {
+              atomicEstimatedRobotPose.set(estimatedRobotPose);
```

Comment:
We discussed this at some point, but do we want to have some kind of `else` when all of these conditions fail? Meaning, if we don't see any targets, or we see them outside of the field, do we want to communicate some kind of failure / no data from vision, rather than report the last known position which could be stale?